uaclient/cli.py

```diff
@@ -29,140 +19,221 @@
 
 DEFAULT_LOG_FORMAT = (
-    '%(asctime)s - %(filename)s:(%(lineno)d) [%(levelname)s]: %(message)s')
+    "%(asctime)s - %(filename)s:(%(lineno)d) [%(levelname)s]: %(message)s"
+)
+
+STATUS_FORMATS = ["tabular", "json"]
 
-STATUS_FORMATS = ['tabular', 'json']
 
+def assert_root(f):
+    """Decorator asserting root user"""
 
-def assert_attached_root(func):
-    """Decorator asserting root user and attached config."""
-    def wrapper(args, cfg):
+    @wraps(f)
+    def new_f(*args, **kwargs):
         if os.getuid() != 0:
             raise exceptions.NonRootUserError()
-        if not cfg.is_attached:
-            raise exceptions.UnattachedError()
-        return func(args, cfg)
+        return f(*args, **kwargs)
+
+    return new_f
+
+
+def assert_attached(unattached_msg_tmpl=None):
+    """Decorator asserting attached config.
+
+    :param unattached_msg_tmpl: Optional msg template to format if raising an
+        UnattachedError
+    """
+
+    def wrapper(f):
+        @wraps(f)
+        def new_f(args, cfg):
+            if not cfg.is_attached:
+                if unattached_msg_tmpl:
+                    name = getattr(args, "name", "None")
+                    msg = unattached_msg_tmpl.format(name=name)
+                    exception = exceptions.UnattachedError(msg)
+                else:
+                    exception = exceptions.UnattachedError()
+                raise exception
+            return f(args, cfg)
+
+        return new_f
+
     return wrapper
 
 
-def attach_parser(parser=None):
-    """Build or extend an arg parser for attach subcommand."""
-    usage = USAGE_TMPL.format(name=NAME, command='attach [token]')
+def require_valid_entitlement_name(operation: str):
+    """Decorator ensuring that args.name is a valid service.
+
+    :param operation: the operation name to use in error messages
+    """
+
+    def wrapper(f):
+        @wraps(f)
+        def new_f(args, cfg):
+            if hasattr(args, "name"):
+                name = args.name
+                tmpl = ua_status.MESSAGE_INVALID_SERVICE_OP_FAILURE_TMPL
+                if name not in entitlements.ENTITLEMENT_CLASS_BY_NAME:
+                    raise exceptions.UserFacingError(
+                        tmpl.format(operation=operation, name=name)
+                    )
+            return f(args, cfg)
+
+        return new_f
+
+    return wrapper
+
+
+def attach_premium_parser(parser):
+    """Build or extend an arg parser for init-attach subcommand."""
+    usage = USAGE_TMPL.format(name=NAME, command="attach-premium")
     if not parser:
         parser = argparse.ArgumentParser(
-            prog='attach',
-            description=('Attach this machine to an existing Ubuntu Advantage'
-                         ' support subscription'),
-            usage=usage)
+            prog="attach-premium",
+            description=(
+                "For premium Ubuntu images, automatically attach"
+                " Ubuntu Advantage support subscription"
```

Comment:
Update description to: "Automatically enable Ubuntu Advantage on a premium Ubuntu image"

Reply:
done